### ISSUE TITLE
Custom icon keeps disappearing

### ISSUE THREAD

**Issue by bendehghan** [dmarmor/epichrome#290]
I'm on 2.4 beta 9.

I have an app with a custom icon but sometimes (many times, most times) the icon is generic chrome. Is that a known issue?

Thanks,
Ben

**Comment by dmarmor** [comment]
Sorry for the slow reply. Are you still having this problem? I've not experienced it myself, nor have I had other user reports on 2.4.0b9. My first suggestion would be to reboot and see if that fixed it. There's always a chance for the system to get confused when dealing with Chrome-based apps, as when they run, by necessity they have the same ID as the real Chrome. If a reboot didn't help and you're still having the problem, let me know and we can try some other things to diagnose. Thanks!

**Comment by ylluminate** [comment]
This keeps happening as well for me in 2.3.x. After a computer reboot or killing all bash instances (since I use zsh normally, thus I know all bash sessions are essentially Epichrome or other ancillary Chrom* related processes generally) Epichrome icons and apps seem to give the proper icon to the OS again for rendering in the Dock and task switcher.

**Comment by talleux** [comment]
Hello @dmarmor 

since a while i have some apps loosing their icon. Using 2.4.0b10
The strange thing is: when i edit the app with the epichrome wizard, re-assign the same or another icon, then the wizard shows me the right icon. After editing i see for instance this:

![image](https://user-images.githubusercontent.com/5243081/107026247-babd4700-67aa-11eb-878f-9d930c934eee.png)

But in the finder or the doc, the icon used is this:

![image](https://user-images.githubusercontent.com/5243081/107026398-efc99980-67aa-11eb-8835-004d2e7c86cd.png)

I tried a lot of icon files, that doesn't seem to be the problem.
I can also assign the icon to the app via the finder (CMD+I on the app-Folder) which stays until the next update of epichrome, as expected.

I used to work well for a while so my modest guess is that it is since i updated Catalina to the latest version 10.15.7 but this is only a guess.

I can see the correct icon in:
- /Applications/Kino.app/Contents/Resources/app.icns
- /Applications/Kino.app/Contents/Resources/Welcome/img/app_icon.png
- /Users/<USERNAME>/Library/Application Support/Epichrome/Apps/Kino790/Welcome/img/app_icon.png

**Comment by dmarmor** [comment]
Hi @talleux,

I'm very confused by this! Most reports of this I've gotten have said the app icon reverts to the Chrome icon. This is the first report I've seen of a completely generic app icon. Could I have a bit more info?

1. What OS version are you on?

2. If you right-click that dock icon and choose `Options->Show in Finder` does it point to the correct app bundle? And does the app have the custom icon or a generic one?

Thanks!

**Comment by talleux** [comment]
Hi @dmarmor 

1) As i said,  i recently updated Catalina to the latest version 10.15.7 (from one version before)
2) Yes, i'm sure. What i can add: when the app isn't started, the these icons are the expected ones:
`
/Applications/Kino.app/Contents/Resources/app.icns
/Applications/Kino.app/Contents/Resources/Welcome/img/app_icon.png
/Users/[USER]/Library/Application Support/Epichrome/Apps/Kino/Welcome/img/app_icon.png
`

When i start the app, this changes to:
`
/Applications/Kino.app/Contents/Resources/app.icns ==> Turns into the Google Chrome Icon!
/Applications/Kino.app/Contents/Resources/Welcome/img/app_icon.png ==> This folder doesn't exist anymore
/Users/[USER]/Library/Application Support/Epichrome/Apps/Kino/Welcome/img/app_icon.png ==> still correct
`

If i can give more information, pls tell me.

**Comment by dmarmor** [comment]
So when the app is not running, and you look at it in the Finder, it shows a generic icon, even though the app.icns file in the bundle's Resources directory is the correct one? That would indicate to me some problem with launch services registration or something. Does a reboot help the situation at all?

As for when the app is running, the fact that the icon changes to the Chrome icon is a problem! That should not be happening. My suggestion there would be to try blowing away the engine and letting Epichrome recreate it. The way to do this is to make sure the app is *not* running, and then delete the folder `/Applications/Epichrome Beta/Payload.noindex/<username>/Kino/Engine`. The next time you run the app, it will recreate the Chrome engine, and hopefully the icon will be correct.

Could you let me know the answer to those questions, and whether recreating the engine helps? Thanks!

**Comment by ylluminate** [comment]
Another peculiar observation is that each time I have seen the Google Chrome icon used instead of the custom Epichrome icon, and if there is an update available: When I update to the new Epichrome (2.3.24 > 2.3.25 for example) the correct icon begins to render again in the Dock and cmd-tab app switcher.

**Comment by dmarmor** [comment]
That's interesting. Have you noted whether any of this coincides with Google Chrome updating itself? I've found Epichrome apps can have odd behavior when Chrome is in the midst of a self-update, or has been updated while an Epichrome app is running. Often in that situation, blowing away the engine (as I outlined above) is the only fix. And when an app updates itself, it always recreates the engine, so that would support that theory...

**Comment by ylluminate** [comment]
In addition to losing icons, at some point in this process the custom Epichrome app name is also lost and `kill -9`'ing bash doesn't resolve either that issue or the icon being gone. It seems that only a system restart resolves it. There must be something else at play that needs to be reset/`kill -HUP`'d that I'm missing...

**Comment by dmarmor** [comment]
The loss of the custom app name (I assume you mean in the menu bar?) also points to a problem with the engine as far as I can tell. If this is happening repeatedly, even after you've tried my suggestion of blowing away the engine and letting the app re-create it, then something weird is definitely going on on your system. If that's the case, please let me know. I may need to get a zip of your app or look at some logs to see if it'll help me diagnose.

**Comment by ylluminate** [comment]
In my case: both Menubar name + cmd-tab task switcher and Dock name.  The name reappears after reboot, so it's limited to a "runtime" issue...

So to be clear: this isn't something that seems to be expressed in the bundle, but rather during runtime and with a reboot helping it, there must be a process I'm not killing or some kind of `tmp` file/folder that's not being cleared out properly...

**Comment by dmarmor** [comment]
Well, that's extremely odd. I've been utterly unable to recreate this on my end. If you do find any process or file that reliably causes/fixes it, please do let me know. What OS version are you on?

**Comment by ylluminate** [comment]
Another odd behavior that seems to continue to occur is that after a new update for Epichrome becomes available and offers the download option at startup of an Epichrome app - and if I don't update - the custom icon no longer is rendered anywhere on the live instance (ie, Dock, cmd-tab). After update the custom icon seems to come back though as previously noted...

**Comment by zmanji** [comment]
I want to share that I had the same problem with the new update for epichrome. These are the steps I took

1. Install epichrome `2.3.25`
2. Create an app with a custom icon
3. Use the app, close it, open it, close it again, see prompt for epichrome update, close app, open app again
4. Observe that in the Dock there is the generic icon.
5. At this point, you can repeat step 3 many times and never see the custom icon in the Dock.
6. Install latest epichrome, in my case `2.3.26`
7. Update the app
8. Observe that the custom icon is now in the Dock

Please let me know if you need more information, I can reproduce this pretty reliability on my laptop, I am running macOS `11.2.1`.

**Comment by dmarmor** [comment]
Yeah, this is very odd behavior. It might be something to do with Big Sur, which 2.3.x is not at all tested on. I'm close to release for 2.4.0, which has a different architecture and might solve this. Meantime, sorry for the trouble. Hopefully the updates will keep fixing it temporarily...

**Comment by ylluminate** [comment]
FYI, this is still happening in 2.4.0.

**Comment by zmanji** [comment]
I repeated my reproducer from above between `2.4.1` and `2.4.2` and I could not get it to occur again. It's possible this issue has been fixed now.

**Comment by dmarmor** [comment]
Yeah, 2.4 has a different app architecture from 2.3, so it's very possible whatever was causing it before is no longer happening. I'll leave this open a bit longer in case it recurs.

**Comment by Madd0g** [comment]
I'm on 2.4.4 (mac, using chrome) and having both mentioned issues:

* icon cross contaminated my chrome (the main browser got the custom app icon), so I changed it back to non-custom
* app name got reset at some point to Chrome too

funny observation is, it's still "reacting" to the old name, I have code that focuses apps based on name and it still works

**Comment by talleux** [comment]
i can confirm this, see #328

**Comment by Madd0g** [comment]
Two things:

1. After I deleted that folder, the correct app name was restored (yay!)
2. Regarding the icon - I think I figured it out, I'm using a custom app switcher replacement called [hyperswitch](https://bahoom.com/hyperswitch) and I'm guessing since the bundle id is of the app is the same as native chrome - the switcher just gets confused and shows the cached icon. Same thing happens with hammerspoon - it's showing me the custom icon for both now.

I'm guessing there's no way of changing the bundle id when using the chrome engine?

**Comment by dmarmor** [comment]
I'm glad that fixed it! And yes, the custom app switched definitely sounds like the culprit. Unfortunately it's not possible to change the bundle ID of a Chrome-engine app. Brave apps should behave themselves even with the custom app switcher.

**Comment by talleux** [comment]
> Thanks for the reports. Regarding @talleux 's comment in issue #328 ("while an app is running, all of a sudden the icon is set to the chrome default. When i close the app, the icon appears again"), and @Madd0g 's here, can you both try deleting the offending app's payload and see if that fixes the problem? Here's how:
> 
> 1. Make absolutely sure the app is _not_ running.
> 2. Find the app's payload folder. If your app is in `/Applications/Epichrome/Apps/YourApp.app`, then the payload will be in `/Applications/Epichrome/Payload.noindex/<username>/<YourAppID>`.
> 3. Delete that folder and empty trash.
> 
> Next time you launch, the app will rebuild the engine payload before launching. When it then launches, is the icon correct, or is it still reverting to Chrome?

Hi @dmarmor unfortunately this didn't do the trick for me. I did:
- Delete the Paypload forlder
- Run Epichrome Scan (just to be sure)  --> nothing to fix here
- Started the app --> no correct icon
- Edit the app, set the icon autom. via URL
- Started the app --> no correct icon

The icon is missing when the app runs as well as when it is closed. In the edit wizard and the welcome page the icon is correct.

I uploaded the app to your dropbox (just removed the bookmark file). There you can see the correct app.icns and document.icns

**Comment by dmarmor** [comment]
Thanks for sending that. The good news is, I'm seeing the same problem you are (or similar anyway--it's showing up for me with a folder icon, for some reason). It also is not creating a bookmark folder or bookmarks for the welcome page on first run, which I need to look into as well. It's just misbehaving in a number of ways, but looks otherwise totally normal. I'll check it out and see what I can figure out!

**Comment by talleux** [comment]
@dmarmor i found the solution!
I pressed cmd+i on the app, selected/clicked on the "wrong" icon here and pressed the delete key

![image](https://user-images.githubusercontent.com/5243081/116861370-46f0d600-ac03-11eb-94db-09cf2b1da49d.png)

I the past i used to _set_ icons this way (pasting an icon an that place) but i "removed" them for a while, otherwise epichrome itself couldn't set the icon. I wonder why it was "set back" this way on this apps but probably it came through the use of backups when solving other problems.

I makes sense to me. I could tell more on that topic if needed but now i'm in a hurry.

Can you reproduce it also on you computer @dmarmor ?

**Comment by dmarmor** [comment]
Yup, that worked for me too! I'm still puzzled by the lack of bookmarking the welcome page (which is supposed to happen even if the app doesn't display it). Does that happen when you update or edit these apps on your system? Or do the welcome page updates get added to the bookmark folder?

**Comment by talleux** [comment]
hmmm, no i have 3 apps that again start to loose their icon while running. After closing them i can see in the finder how they get back their icon.
Editing the app and setting the icon again helps but how long? One of them was that Toggl App from #328 where i definitely already fixed the icon that way a few days ago...
2 of the apps concerned are quite recent and where created with epichrome > 4.3 i think.

**Comment by ylluminate** [comment]
I too am having the icons disappear. I also am having some Brave app instances lose my saved passwords like the Chrome app instances...

**Comment by talleux** [comment]
i have this now for more and more apps...
Start the app --> icon set to chrome default. I can see this happen in the finder
close the app --> App is correct in the finder

**Comment by talleux** [comment]
oh yes also i have the behaviour that sometimes i'm logged out from all webapps/websites after starting an app
